xrpl4j-client/src/main/java/org/xrpl/xrpl4j/client/XrplClient.java

```diff
@@ -232,6 +239,102 @@ public FeeResult fee() throws JsonRpcClientErrorException {
     return jsonRpcClient.send(request, FeeResult.class);
   }
 
+  /**
+   * Get the ledger index of a tx result response. If not present, throw an exception.
+   *
+   * @return A string containing value of last validated ledger index.
+   */
+  private UnsignedInteger getMostRecentlyValidatedLedgerIndex() throws JsonRpcClientErrorException {
+    JsonRpcRequest request = JsonRpcRequest.builder()
+      .method(XrplMethods.LEDGER)
+      .addParams(LedgerRequestParams.builder().ledgerSpecifier(LedgerSpecifier.VALIDATED).build())
+      .build();
+    return jsonRpcClient.send(request, LedgerResult.class).ledgerIndexSafe().unsignedIntegerValue();
+  }
+
+  /**
+   * Enumeration of the types of responses from isFinal function.
+   */
+  public enum FinalStatus {
+    /**
+     * Transaction has not been validated until this moment, which means it could be validated in the
+     * following closed ledgers, or, the XRPL peer you are connected to could be missing the ledgers
+     * it would have been validated on.
+     */
+    NOT_FINAL,
+    /**
+     * Transaction is final/validated on the ledger successfully.
+     */
+    VALIDATED_SUCCESS,
+    /**
+     * Transaction is final/validated on the ledger with some failure.
+     */
+    VALIDATED_FAILURE,
+    /**
+     * The lastLedgerSequence of the tx has passed, i.e., transaction was not included in any closed
+     * ledgers ranging from the one it was submitted on to the lastLedgerSequence.
+     */
+    EXPIRED
+  }
+
+  /**
+   * Check if the transaction is final on the ledger or not.
+   *
+   * @param transactionHash {@link Hash256} of the submitted transaction to check the status for.
+   * @param submittedOnLedgerIndex {@link LedgerIndex} on which the transaction with hash transactionHash
+   *        was submitted. This can be obtained from submit() response of the tx as validatedLedgerIndex.
+   * @return {@code true} if the {@link Transaction} is final/validated else {@code false}.
+   * @throws JsonRpcClientErrorException if {@code jsonRpcClient} throws an error.
+   */
+  public FinalStatus isFinal(
+    Hash256 transactionHash,
+    LedgerIndex submittedOnLedgerIndex,
+    UnsignedInteger lastLedgerSequence
+  ) throws JsonRpcClientErrorException {
+
+    TransactionResult response = this.transaction(TransactionRequestParams.of(transactionHash), Transaction.class);
```

Comment:
If the transaction doesn't exist on ledger, this line will throw a `JsonRpcClientErrorException`. However, we need to know if the transaction isn't found so we can hit line 311.

I'd suggest adding a method like this:
```java
protected Optional<? extends TransactionResult<? extends Transaction>> getValidatedTransaction(
  final Hash256 transactionHash
) {
  Objects.requireNonNull(transactionHash);

  try {
    final TransactionResult<? extends Transaction> transactionResult = xrplClient.transaction(
      TransactionRequestParams.of(transactionHash), Transaction.class
    );
    return Optional.ofNullable(transactionResult).filter(TransactionResult::validated);
  } catch (JsonRpcClientErrorException e) {
    // The transaction was not found on ledger, warn on this, but otherwise return.
    logger.warn(e.getMessage(), e);
    return Optional.empty();
  }
}
```

Then you can call that method and then do `.map(...).orElseGet(() -> ...)` instead of your `if`/`else`s that you have here.